src/node.cc

```diff
@@ -446,22 +453,50 @@ void LoadEnvironment(Environment* env) {
   USE(StartMainThreadExecution(env));
 }
 
+#if defined __linux__ || defined __APPLE__
+static std::atomic<void (*)(int signo, siginfo_t* info, void* ucontext)>
+    previous_sigsegv_action;
+
+void TrapWebAssemblyOrContinue(int signo, siginfo_t* info, void* ucontext) {
+  if (!v8::TryHandleWebAssemblyTrapPosix(signo, info, ucontext)) {
+    CHECK(previous_sigsegv_action.is_lock_free());
+    auto prev = previous_sigsegv_action.load();
+    if (prev != nullptr) {
+      prev(signo, info, ucontext);
+    } else {
+      uv_tty_reset_mode();
+      raise(signo);
+    }
+  }
+}
+#endif  // defined __linux__ || defined __APPLE__
 
 #ifdef __POSIX__
+
 void RegisterSignalHandler(int signal,
-                           void (*handler)(int signal),
-                           bool reset_handler) {
+                           void (*handler)(int signal,
+                                           siginfo_t* info,
+                                           void* ucontext),
+                           bool reset_handler,
+                           struct sigaction* oldact) {
   struct sigaction sa;
   memset(&sa, 0, sizeof(sa));
-  sa.sa_handler = handler;
+#if defined __linux__ || defined __APPLE__
+  if (signal == SIGSEGV) {
+    sa.sa_sigaction = previous_sigsegv_action.exchange(handler);
+    *oldact = sa;
```

Comment:
This will segfault when `oldact == nullptr`. Callers never pass a non-nullptr pointer anyway, it looks like, so I would just remove the parameter and change the `.exchange()` to a `.store()`.